### ISSUE TITLE
Configuration options in Settings

### ISSUE THREAD

**Issue by maninak** [cytechmobile/radicle-vscode-extension#8]
**path to rad binary**

No need to have a default option (on all platforms). Match default with radicle installation guide. Attempt to run which rad and if found set output as the value of the field. If path hasn't been set we have to tell the user. (Dismiss / Resolve -> takes you takes you to settings OR (better?) asks for input in the command palette (?) )

Once a path has been defined, it would be great if we had a Test button that ensures rad --version can be invoked (so that we have the minimum CLI version installed and path / file exist and are executable, etc. ).

**Comment by maninak** [comment]
In practice, this ticket turned out to be the equivalent work of multiple tickets (as usually specced) in one, due to the branching factor and the numerous VS Code APIs needed to be used for the first time.

This is a rough sketch outlining the missing rad CLI handler's logic:

![Image](https://user-images.githubusercontent.com/7542054/227822096-cba813a5-02c9-4520-8731-30d6bcb8d9c5.jpeg)

Notable features of the solution are:
- `pathToBinary` config is placed under the "Advanced" settings subcategory
- `pathToBinary` config by default isn't synced along with the user's VS Code config across computers and is visually marked as such. However, it can still be synced if the user chooses and is additionally overridable by workspace config if necessary.
- `pathToBinary` config's string input is trimmed
- edits to `pathToBinary` config take effect immediately as the user types (denounced event handler), without requiring IDE restart
- When the extension cannot resolve `rad` the user is notified with a warning explaining the cause of the issue setting UX expectations and offering guided resolution with a multi-branch and multi-step troubleshooter
- The troubleshooter follows all UX best practices defined in the VS Code extension authoring docs.
- The troubleshooter results in either
    1. open the browser to https://radicle.xyz/get-started.html
    2. browse the filesystem using the native file picker
    3. type the path into an InputBox with on-the-fly path validation
- The resolved `rad` version and path are logged upon the extension's activation
- Upon the extension's activation, or upon edits to `pathToBinary` config (via any current or future method), a hook will validate that the `rad` CLI resolves and if not, will notify the user and offer the option to troubleshoot as previously described, always providing timely feedback to the user.
- No `rad` CLI installation validation (and thus notifications) take place, unless the extension is somehow activated (e.g. command is run from Command Pallete, Radicle View Container is shown in the Activity Bar, etc).

A demo showcasing some of the above features:

https://user-images.githubusercontent.com/7542054/227825688-83f1b8de-ffb1-4add-bbb5-40f3a0c38d5e.mp4

**Comment by gsaslis** [comment]
Nice work @maninak - and 👏 on its documentation! 

If I understand correctly, if the path in the settings is **empty**, VS code can still use `rad` from e.g. $PATH ? 

If so, I'm wondering if we should call that setting path **override** or something to indicate what the value there is meant for. 

Or maybe we auto-detect `rad` and we pre-fill that field with the value from `$PATH` ?

**Comment by maninak** [comment]
> If I understand correctly, if the path in the settings is **empty**, VS code can still use `rad` from e.g. $PATH ?

Yes. Following the Getting Started guide, `rad` should be globally accessible in the shell, at least on Linux which I've tested (effectively it's appended in $PATH). So, by default, the extension will just run `rad push`, or `rad pull`.

If the user sets `pathToBinary: '/usr/bin/rad` then the commands shown above will be run in the shell as `/usr/bin/rad push` and `/usr/bin/rad pull`.

The above can be seen in the source code here https://github.com/cytechmobile/radicle-vscode-extension/blob/73a05d985710f0c45ce445a7c900b31c9b245dba/src/utils/radCli.ts#L12

> Or maybe we auto-detect `rad` and we pre-fill that field with the value from `$PATH`?

Nop, we don't pre-fill/set the config `pathToBinary`, because that would mean that we're setting a user config without explicit user action, which as it turns out is bad practice.

If we pre-filled the config after reading its path with something like `which rad` (which sources the response from `$PATH`) and then the user changed the location of the binary and the related definition in $PATH, our extension will fall out of sync and error out, even though the user has done everything right and running `rad --version` in the terminal would still work.

FYI every config that is set adds a new entry in the `settings.json`, which some people maintain directly instead of using the settings UI (has rich autocompletion and JSDoc-powered tooltip docs). Furthermore, uninstalling an extension doesn't remove its configs set in `settings.json` (in case the extension is later re-installed). For those two reasons, we must avoid unnecessarily and unilaterally "littering" the file with configs. And that's why if the setting is "undefined" we default calling the CLI just `rad` which is a sane default per the Getting Started guide (and if that fails, our new user notification and troubleshooting process is available to take up the slack).

**Comment by gsaslis** [comment]
Thanks for the additional context. 

> Following the Getting Started guide, rad should be globally accessible in the shell, at least on Linux which I've tested (effectively it's appended in $PATH). So, by default, the extension will just run rad push, or rad pull.

I see 2 different processes being referenced here: 
- a process spawned in a shell, where $PATH can be read / modified a bunch of different ways - e.g. `.bashrc` , `.bash_profile` (and other terminal variations), etc. : let's say this process is `rad --version`
- a process that is spawned inside VS Code somehow, by our extension, which also eventually invokes `rad --version`

How can we be sure that these 2 processes read the same `$PATH`? 🤔 

If our recommendation is "leave the setting empty and set your $PATH correctly", I think it's important to clarify what the valid ways to set the $PATH are.  

( e.g. from a quick search, I see it has been an [issue](https://stackoverflow.com/questions/43983718/set-global-path-environment-variable-in-vs-code) - but there are some newer developments that mean maybe this [isn't (?) so much of a concern any more](https://code.visualstudio.com/docs/supporting/faq#_resolving-shell-environment-fails)

With that in mind, perhaps it would make sense to add a simple "learn more about how to correctly set $PATH on your platform" in the settings screen ? I am not 100% sure that the link above is the best page to link to though... 

what do you think?

**Comment by maninak** [comment]
Hey, thanks for describing your thoughts in detail and for providing the references.

Personally, I've never experienced any `$PATH`- or shell-related issues with either VS Code or Linux and I've been using both explicitly since 2016 and 2009 respectively.

IMO I think this whole topic rad CLI resolution topic is pretty simple:

1. for most people, in most cases everything will "just work"™
2. if not, for whatever reason, all the user has to do is download the `rad` binary and save it anywhere on their machine, and via any of the 3 available methods we offer, set the `pathToBinary` config in the settings.

There's no need for them to fiddle with advanced `$PATH` stuff or shell config like `.bashrc` (although if they do that, our extension will again work 🙂).

------------

So, to recap, as we have things now, everything will either just work, or _at worst_ the user will download a file and tell the extension where it is.

That's pretty good UX, IMO, and could only get better if we bundled rad into the extension itself (which requires plenty of CI and build-tooling work but is indeed doable).

**Comment by gsaslis** [comment]
ok, I think you're right and that makes sense.

Adding $PATH into the equation probably confuses things more rather than asking them to point to the binary, with the guide you've already created here. 

All good - thanks for the input!

**Comment by maninak** [comment]
> Adding $PATH into the equation probably confuses things more rather than asking them to point to the binary, with the guide you've already created here.

Yup, that's why I only use this more advanced terminology under the "Advanced" config's description in the settings and nowhere else; neither in the notification nor in any copy throughout the troubleshooting flow. That's on purpose, not by accident. I spend a lot of time crafting those copies (not implying they are perfect and shouldn't be changed, though).

> All good - thanks for the input!

Sure thing! :beers:
